File: weread_2_notion.py
import time

from pywebio.output import put_text, put_info

from logger import info, debug, warning, error, exception
from notion.notion import NotionClient
from weread.weread import WeRead

try:
    from settings.settings import WEREAD_COOKIE, DATABASE_ID, BOOK_BLACKLIST, NOTION_TOKEN
except Exception as e:
    error(f'导入配置文件失败，{e} Web 无伤大雅')


def weread_2_notion(notion_token=NOTION_TOKEN,
                    weread_cookie=WEREAD_COOKIE,
                    database_id=DATABASE_ID,
                    book_blacklist=BOOK_BLACKLIST,
                    is_web=True):
    try:
        weread_ = WeRead(weread_cookie)
    except Exception as e:
        raise Exception(f'微信读书登录失败，{e}')
    try:
        notion = NotionClient(notion_token, database_id)
    except Exception as e:
        raise Exception(f'notion 登录失败，{e}')
    # 书籍列表
    try:
        books = weread_.get_notebooklist()
    except Exception as e:
        raise Exception(f'获取书籍列表失败，请检查微信读书 cookies 是否正确 {e.__str__()}')
    all_book = []
    handled_book = []
    ignore_book = []
    err_book = {}
    info(f'书籍目录：{books}')
    if books is not None:
        for book in books:
            try:
                sort = book["sort"]  # 更新时间
                book = book.get("book")
                title = book.get("title")
                info(f' 正在同步：《{title}》, 当前进度 {len(handled_book) + len(ignore_book)}/{len(books)}')
                if is_web:
                    put_text(f' 正在同步：《{title}》, 当前进度 {len(handled_book) + len(ignore_book)}/{len(books)}')
                all_book.append(title)
                if title in book_blacklist:
                    info(f'《{title}》在黑名单中，跳过')
                    ignore_book.append(title)
                    continue
                # 调试用途代码
                # if book.get("title") != '黄金时代':
                #     ignore_book.append(title)
                #     continue
                # print(book['123'])
                # print(book)
                if sort <= notion.get_sort():
                    warning(f'当前图书《{title}》没有更新划线、书评等信息，暂不处理')
                    ignore_book.append(title)
                    continue
                handled_book.append(title)
                cover = book.get("cover", '没有封面')
                bookId = book.get("bookId", '-1')
                author = book.get("author", '没有作者信息')
                info(f'开始处理《{title}》, bookId={bookId}, sort={sort}')
                notion.check(bookId)  # TODO 如果自行在 notion 修改，这里会删除重新插入，规避这个逻辑
                chapter = weread_.get_chapter_info(bookId)
                bookmark_list = weread_.get_bookmark_list(bookId)
                summary, reviews = weread_.get_review_list(bookId)
                bookmark_list.extend(reviews)
                bookmark_list = sorted(bookmark_list, key=lambda x: (
                    x.get("chapterUid", 1),
                    0 if (x.get("range", "") == "" or x.get("range").split("-")[0] == "") else int(
                        x.get("range").split("-")[0])))
                isbn, rating, intro, category = weread_.get_bookinfo(bookId)
                children, grandchild = notion.get_children(
                    chapter, summary, bookmark_list)
                block_id = notion.insert_to_notion(bookName=title,
                                                   bookId=bookId,
                                                   book_str_id=weread_.calculate_book_str_id(bookId),
                                                   cover=cover,
                                                   sort=sort,
                                                   author=author,
                                                   isbn=isbn,
                                                   rating=rating,
                                                   intro=intro,
                                                   category=category)
                results = notion.add_children(block_id, children)
                if (len(grandchild) > 0 and results != None):
                    notion.add_grandchild(grandchild, results)
                debug(f'结束处理《{title}》, bookId={bookId}, sort={sort}')
            except Exception as e:
                exception(f'处理书籍《{title}》失败')
                err_book[title] = [str(e), book]
                continue
        return all_book, handled_book, ignore_book, err_book


if __name__ == '__main__':
    weread_2_notion(is_web=False)
    # weread = WeRead(WEREAD_COOKIE)
    # print(weread.session.cookies.get_dict())
    # res = weread.session.get(
    #     'https://i.weread.qq.com/users/info')
    # weread.session.
    # print(weread.get_notebooklist())

File: plugin/weread_2_notion.js
// const fetch = require('node-fetch');
const fetch = require('isomorphic-fetch');
const {WeRead} = require('./weread.js');
const {NotionClient} = require('./notion');

const WEREAD_COOKIE = 'uin=o1307317886; skey=@Q8MrVr0ER; RK=Dn35RmM4ef; ptcz=bc8fbc0bf0c2ff9ffa6fc1fccc5209246c7dc595183dfadfcbec558cb8560c2d; pgv_info=ssid=s5596440080; pgv_pvid=1180932580; midas_openid=89C1944462BF4A87C4C951878B816A12; midas_openkey=5BBE8339ECB7A44878A632E5D7ABE8AE; wr_fp=2533447432; wr_gid=262277338; pac_uid=1_1307317886; wr_vid=23683664; wr_pf=0; wr_rt=web%40AVEURyUiB24N6OmM1Q1_AL; wr_localvid=fb3325d071696250fb37ae2; wr_name=%E8%AF%B6%E9%B8%AD; wr_gender=1; wr_skey=zOSz_VvI; wr_avatar=https%3A%2F%2Fthirdwx.qlogo.cn%2Fmmopen%2Fvi_32%2FXpfWjDr8uAYiagxaSibFmnMh2LUJXTQrjJ8z2ve6X8J3w2CI79YSvQwic1icyHUoHVgoccckuoLxlt1cQOD87tZLfQ%2F132';
const DATABASE_ID = 'd92bb4b8434745baa2061caf67d6ef7a';
const BOOK_BLACKLIST = ['从红月开始',]; // Array of book titles to ignore
const NOTION_TOKEN = 'secret_kZMZkH95SLspBpZb0OPD49pTZ2JuGL2z9DeyoHpe3DG';

async function weread2notion() {
    const weread = new WeRead(WEREAD_COOKIE);
    const notion = new NotionClient(NOTION_TOKEN, DATABASE_ID);

    // Get the book list
    const books = await weread.get_notebooklist();

    if (books !== null) {
        for (const book of books) {
            const sort = book.sort; // Update time
            const {book: bookData} = book;
            const title = bookData.title;

            if (BOOK_BLACKLIST.includes(title)) {
                console.log(`《${title}》 is in the blacklist, skipping`);
                continue;
            }

            if (sort <= notion.getSort()) {
                console.log(`Current book 《${title}》 does not have updated underlines, reviews, etc., skipping`);
                continue;
            }

            const cover = bookData.cover;
            const bookId = bookData.bookId;
            const author = bookData.author;

            console.log(`Processing 《${title}》, bookId=${bookId}, sort=${sort}`);

            await notion.check(bookId); // TODO: If modified manually in Notion, this will delete and reinsert, avoid this logic

            const chapter = await weread.getChapterInfo(bookId);
            const bookmarkList = await weread.getBookmarkList(bookId);
            const {summary, reviews} = await weread.getReviewList(bookId);

            bookmarkList.push(...reviews);
            bookmarkList.sort((a, b) => {
                const aChapterUid = a.chapterUid || 1;
                const bChapterUid = b.chapterUid || 1;
                const aRange = a.range || '';
                const bRange = b.range || '';
                const aStart = aRange !== '' ? parseInt(aRange.split('-')[0]) : 0;
                const bStart = bRange !== '' ? parseInt(bRange.split('-')[0]) : 0;
                return aChapterUid - bChapterUid || aStart - bStart;
            });

            const {isbn, rating} = await weread.getBookInfo(bookId);
            const {children, grandchild} = await notion.getChildren(chapter, summary, bookmarkList);

            const blockId = await notion.insertToNotion({
                bookName: title,
                bookId,
                bookStrId: weread.calculateBookStrId(bookId),
                cover,
                sort,
                author,
                isbn,
                rating
            });

            const results = await notion.addChildren(blockId, children);

            if (grandchild.length > 0 && results !== null) {
                await notion.addGrandchild(grandchild, results);
            }

            console.log(`Finished processing 《${title}》, bookId=${bookId}, sort=${sort}`);
        }
    }
}

weread2notion().catch((error) => {
    console.log('An error occurred:', error);
});
